### PR TITLE
Add charts

### PR DESCRIPTION
**Pull request by LadyPython**

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/data/User.java [hunk 1]
```diff
@@ -30,4 +30,14 @@ public static String getUserNickname(String id) {
...
   }
+  public static Integer getVote(String id) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query("Vote").setFilter(new Query.FilterPredicate("id", Query.FilterOperator.EQUAL, id));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      return null;
+    }
+    return (int) (long) entity.getProperty("slice");
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/data/User.java:41]: Why are you converting to first to long and then to int? 
Maybe you can also use Integer.parseInt to catch conversion errors?

File: portfolio/src/main/java/com/google/sps/servlets/PizzaSliceDataServlet.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    pizzaVotes.put(0, 0);
+    pizzaVotes.put(1, 0);
+    pizzaVotes.put(2, 0);
+    pizzaVotes.put(3, 0);
+    pizzaVotes.put(4, 0);
+    pizzaVotes.put(5, 0);
+    for (Entity entity : results.asIterable()) {
+      Integer slice = (int) (long) entity.getProperty("slice");
+      pizzaVotes.put(slice, pizzaVotes.get(slice) + 1);
+    }
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/PizzaSliceDataServlet.java:58]: This looks dangerous. If there is a slice value in one entity that is not in [0,...,5] you get a null pointer back from the get() method.
Maybe better use https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#getOrDefault-java.lang.Object-V- with default 0 and remove the lines where you initialize the hash map. 

Another question: Why are you using a hashmap instead of a simple array if you expect the slice values to be in [0,..,5]?

File: portfolio/src/main/java/com/google/sps/servlets/PizzaSliceDataServlet.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+      return;
+    }
+    
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity entity = new Entity("Vote", id);
+    entity.setProperty("id", id);
+    entity.setProperty("slice", slice);
+    datastore.put(entity);
+
+    response.sendRedirect("/tmp.html");
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/PizzaSliceDataServlet.java:102]: Maybe use a better name than tmp.html :)

File: portfolio/src/main/java/com/google/sps/servlets/UserInfoServlet.java [hunk 4]
```diff
@@ -60,16 +60,18 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     response.setContentType("application/json;");
     UserService userService = UserServiceFactory.getUserService();
     
+    String destinationURL = request.getHeader("referer");
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/UserInfoServlet.java:63]: What if there is no such header?

File: portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Comparator;
+import java.util.Scanner;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/vk-data")
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java:31]: What does vk stands for? Try to avoid abbreviations in names if not clear from the context.

File: portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        break;
+      case "cities":
+        json = gson.toJson(cities_sorted);
+        break;
+      case "genders":
+        json = gson.toJson(genders);
+        break;
+      case "ages":
+        json = gson.toJson(views);
+        break;
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java:158]: Please also add a default case.

File: portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+  @Override
+  public void init() {
+    Map<String, Integer> countries = new HashMap<>();
+    Map<String, Integer> cities = new HashMap<>();
+
+    Scanner scanner = new Scanner(getServletContext().getResourceAsStream(
+        "/WEB-INF/vkontakte_stat.csv"));
+    while (scanner.hasNextLine()) {
+      String line = scanner.nextLine();
+      String[] cells = line.split(",(?=([^\"]*\"[^\"]*\")*[^\"]*$)");
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java:48]: Can you extract ",(?=([^\"]*\"[^\"]*\")*[^\"]*$)" to a const value in the class and give it a readable name? It is hard to parse what you are doing here.

File: portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+      String year = cells[0].replace("\"", "").split("\\.")[2];
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java:49]: You are doing .replace("\"", "") all over the place, maybe create a static function called removeQuotes?
Maybe even call it on cells since you are remove quotes everywhere.

File: portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+          try {
+            Integer value = Integer.valueOf(cells[4]);
+            Integer current_value = countries.get(country);
+            if (!country.equals("Russia") && !country.equals("")) {
+              countries.put(country, current_value == null ? value : current_value + value);
+            }
+          } catch (NumberFormatException e) {
+            System.out.println(e);
+          }
+          break;
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java:63]: You can simplify this switch/case statement:
Do the key/value extraction before:
```
String key = cells[2];
if(key.isEmpty()) {continue;}
Integer value = 0;
try {
value = Integer.valueOf(cells[4]);
} catch() {
...
continue;
}
if(value == 0) { continue; }
```

Then a simple case statement:
```
switch() {
case "countries":
   countries.put(key, countries.getOrDefault(key, 0) + value);
}
```

File: portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+            }
+            if (value != 0) {
+              Integer current_value = views_for_year.get("total");
+              views_for_year.put("total", current_value == null ? value : current_value + value);
+              views.put(year, views_for_year);
+            }
+          } catch (NumberFormatException e) {
+            System.out.println(e);
+          }
+          break;
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java:106]: This one can then be simplified to:
```
Map<String, Integer> views_for_year = views.get(year);
if (views_for_year == null) {
    views_for_year = new HashMap<String, Integer>();
    views.put(year, views_for_year);
}
views_for_year.put("total", views_for_year.getOrDefault("total", 0) + value);
```

File: portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+            } else {
+                views_for_year = views.get(year);
+            }
+            Integer current_value = views_for_year.get(age);
+            views_for_year.put(age, current_value == null ? value : current_value + value);
+            views.put(year, views_for_year);
+          } catch (NumberFormatException e) {
+            System.out.println(e);
+          }
+          break;
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java:123]: Similar simplifications as for "views".

File: portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+      }
+    }
+    scanner.close();
+
+    List<Map.Entry<String, Integer>> countries_list = new ArrayList<>(countries.entrySet());
+    countries_list.sort(Map.Entry.comparingByValue(Comparator.reverseOrder()));
+    for (Map.Entry<String, Integer> entry : countries_list) {
+        countries_sorted.put(entry.getKey(), entry.getValue());
+    }
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java:132]: Maybe this is a good example to start using Java streams :) 
https://www.baeldung.com/java-hashmap-sort#value_5
There is also an example how to get the sorted linkedhashmap.

Since you are doing the same twice, please use a method instead of doubling the code.

File: portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java (hunk 5 above)
Comment (LadyPython) [reply portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java:31]: It is name for russian "Facebook" :)

File: portfolio/src/main/java/com/google/sps/servlets/PizzaSliceDataServlet.java (hunk 2 above)
Comment (LadyPython) [reply portfolio/src/main/java/com/google/sps/servlets/PizzaSliceDataServlet.java:58]: Yes, it's the good question why -- I'll change. 
What about entity: could I somehow guarantee that values is in [0, 5]? When I put in to datastore, for example. Or should I also check here that the value is in interval?

Comment (marcbury-g) [reply portfolio/src/main/java/com/google/sps/servlets/PizzaSliceDataServlet.java:58]: You should check it here too. Maybe you change the code for putting it into the datastore or there is a bug in that code then it would also break this here.

File: portfolio/src/main/java/com/google/sps/servlets/PizzaSliceDataServlet.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    PreparedQuery results = datastore.prepare(query);
+
+    List<Integer> pizzaVotes = new ArrayList<>(Arrays.asList(0, 0, 0, 0, 0, 0));
+    for (Entity entity : results.asIterable()) {
+      Integer slice = Integer.parseInt(entity.getProperty("slice").toString());
+      if (slice > 5 || slice < 0) {
+        System.err.println("Slice number is not in [0, 5]: " + slice);
+        response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+        response.getWriter().println("Slice number should be in [0, 5]");
+        return;
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/PizzaSliceDataServlet.java:55]: I would not respond with BAD_REQUEST here because the data is coming from internal storage.  BAD_REQUEST means that the client did something wrong which is not the case here.
Maybe better just do the error print and ignore this value.

File: portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+      catch (NumberFormatException e) {
+        System.out.println(e);
+        continue;
+      }
+
+      Map<String, Integer> views_for_year = views.getOrDefault(year, new HashMap<String, Integer>());
+
+      switch (type) {
+        case "countries":
+          if (!key.equals("Russia") && !key.equals("")) {
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java:72]: You are already checking for empty key in line 55.

File: portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+            countries.put(key, countries.getOrDefault(key, 0) + value);
+          }
+          break;
+        case "cities":
+          if (!key.equals("")) {
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java:77]: You are already checking for empty key in line 55.

File: portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+      case "cities":
+        json = gson.toJson(cities);
+        break;
+      case "genders":
+        json = gson.toJson(genders);
+        break;
+      case "ages":
+        json = gson.toJson(views);
+        break;
+      default:
```
Comment (marcbury-g) [inline portfolio/src/main/java/com/google/sps/servlets/VKDataServlet.java:120]: Maybe return BAD_REQUEST?

File: portfolio/src/main/java/com/google/sps/servlets/PizzaSliceDataServlet.java (hunk 13 above)
Comment (pmueller-gg) [reply portfolio/src/main/java/com/google/sps/servlets/PizzaSliceDataServlet.java:55]: Wouldn't it be SC_INTERNAL_SERVER_ERROR in that case?
